src/nbody_simulation_result.jl

```diff
@@ -124,30 +265,52 @@ end
 
 function initial_energy(simulation::NBodySimulation)
     (u0, v0, n) = gather_bodies_initial_coordinates(simulation.system)
-    return potential_energy(u0, simulation) + kinetic_energy(v0, simulation) 
+    ms = get_masses(simulation.system)
+    return potential_energy(u0, simulation) + kinetic_energy(v0, ms) 
 end
 
 # Instead of treating NBodySimulation as a DiffEq problem and passing it into a solve method
 # it is better to use a specific function for n-body simulations.
 function run_simulation(s::NBodySimulation, alg_type=Tsit5(), args...; kwargs...)
-    initial_en = initial_energy(s)
-    function energy_manifold!(residual, u)
-        n = length(s.system.bodies)
-        vs = @view u[:, n+1:end]
-        us = @view u[:,1:n]
-        residual[:,n+1:end] = initial_en - kinetic_energy(vs, s) - potential_energy(us, s)
-    end
-    energy_cb = ManifoldProjection(energy_manifold!)
     solution = solve(ODEProblem(s), alg_type, args...; kwargs...)
     return SimulationResult(solution, s)
 end
 
 # this should be a method for integrators designed for the SecondOrderODEProblem (It is worth somehow to sort them from other algorithms)
 function run_simulation(s::NBodySimulation, alg_type::Union{VelocityVerlet,DPRKN6,Yoshida6}, args...; kwargs...)
-    solution = solve(SecondOrderODEProblem(s), alg_type, args...; kwargs...)
+    cb = obtain_callbacks_for_so_ode_problem(s)
+    solution = solve(SecondOrderODEProblem(s), alg_type, args...; callback=cb, kwargs...)
     return SimulationResult(solution, s)
 end
 
+function obtain_callbacks_for_so_ode_problem(s::NBodySimulation)
+    callback_array = Vector{DECallback}()
+
+    if s.thermostat isa AndersenThermostat
+        push!(callback_array, get_andersen_thermostating_callback(s))
+    end
+
+    return CallbackSet(tuple(callback_array...)...)
+end
+
+function get_andersen_thermostating_callback(s::NBodySimulation)
+    p = s.thermostat::AndersenThermostat
+    n = length(s.system.bodies)
+    v_dev = sqrt(p.kb * p.T / s.system.bodies[1].m)
+
+    condition = function (u, t, integrator)
+        true
+    end
+    affect! = function (integrator)
+        for i = 1:n
+            if randn() < p.ν * (integrator.t - integrator.tprev)
+                integrator.u.x[1][:,i] .= v_dev * randn(3)
```

Comment:
This'll be a bunch of allocations. You'll want to broadcast instead of `randn(3)`